pkg/autoupdate/tuf/autoupdate.go

```diff
@@ -0,0 +1,231 @@
+	if err := os.MkdirAll(localTufDirectory, 0750); err != nil {
+		return nil, fmt.Errorf("could not make local TUF directory %s: %w", localTufDirectory, err)
+	}
+
+	// Set up our local store i.e. point to the directory in our filesystem
+	localStore, err := filejsonstore.NewFileJSONStore(localTufDirectory)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize local TUF store: %w", err)
+	}
+
+	// Set up our remote store i.e. tuf-devel.kolide.com
+	remoteOpts := client.HTTPRemoteOptions{
+		MetadataPath: "/repository",
+	}
+	remoteStore, err := client.HTTPRemoteStore(metadataUrl, &remoteOpts, metadataHttpClient)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize remote TUF store: %w", err)
+	}
+
+	metadataClient := client.NewClient(localStore, remoteStore)
+	if err := metadataClient.Init(rootJson); err != nil {
+		return nil, fmt.Errorf("failed to initialize TUF client with root JSON: %w", err)
+	}
+
+	return metadataClient, nil
+}
+
+func (ta *TufAutoupdater) Execute() (err error) {
+	checkTicker := time.NewTicker(ta.checkInterval)
+	errorCheckTicker := time.NewTicker(1 * time.Hour)
+	cleanupTicker := time.NewTicker(12 * time.Hour)
+
+	for {
+		select {
+		case <-checkTicker.C:
+			if err := ta.checkForUpdate(); err != nil {
+				ta.lock.Lock()
+				ta.errorCounter = append(ta.errorCounter, time.Now().Unix())
+				ta.lock.Unlock()
+
+				level.Debug(ta.logger).Log("msg", "error checking for update", "err", err)
+			}
+		case <-errorCheckTicker.C:
+			rollingErrorCount := ta.rollingErrorCount()
+			level.Debug(ta.logger).Log("msg", "checked rolling error count for TUF updater", "err_count", rollingErrorCount, "binary", ta.binary)
+		case <-cleanupTicker.C:
+			ta.cleanUpOldErrorCounts()
+		case <-ta.interrupt:
+			level.Debug(ta.logger).Log("msg", "received interrupt, stopping")
+			return nil
+		}
+	}
+}
+
+func (ta *TufAutoupdater) Interrupt(_ error) {
+	ta.interrupt <- struct{}{}
+}
+
+func (ta *TufAutoupdater) checkForUpdate() error {
+	_, err := ta.metadataClient.Update()
+	if err != nil {
+		return fmt.Errorf("could not update metadata: %w", err)
+	}
+
+	// Find the newest release for our channel -- right now for logging purposes only
+	targets, err := ta.metadataClient.Targets()
+	if err != nil {
+		return fmt.Errorf("could not get complete list of targets: %w", err)
+	}
+
+	targetReleaseFile := fmt.Sprintf("%s/%s/%s/release.json", ta.binary, ta.operatingSystem, ta.channel)
+	for targetName, target := range targets {
+		if targetName != targetReleaseFile {
+			continue
+		}
+
+		// We found the release file that matches our OS and binary. Evaluate it
+		// to see if we're on this latest version.
+		type releaseFileCustomMetadata struct {
+			Target string `json:"target"`
+		}
+
+		var custom releaseFileCustomMetadata
+		if err := json.Unmarshal(*target.Custom, &custom); err != nil {
+			return fmt.Errorf("could not unmarshal release file custom metadata: %w", err)
+		}
+
+		level.Debug(ta.logger).Log(
```

Comment:
just to make sure I understand, this is all we are doing for now? logging when we find a newer version?

Reply:
Yes correct! Next steps (in next PRs) will be a) expose this info in a launcher table so we can query and see that devices are correctly pulling from new TUF and then b) start to actually download these newer versions.